### ISSUE TITLE
Accessing a detroyed object in SceneManager::processStateChanges()

### ISSUE THREAD

**Issue by athanase** [GameFoundry/bsf#173]
Hi,
I am using the lastest version of `bsf` with a custom version of the Banshee3D editor.
I recently ran into an error after pulling the latest `bsf` changes, and I think it has to do with the new `SceneManager` logic (not sure though). The application crashes after closing a menu for example.

I will try and get a closer look later. Any idea what might be happenning ?

Here is the crash log:
```
A fatal error occurred and the program has to terminate!
  - Error: InternalErrorException
  - Description: Trying to access an object that has been destroyed.
  - In function: void bs::GameObjectHandleBase::throwIfDestroyed() const
  - In file: editor/Source/bsf/Source/Foundation/bsfCore/Scene/BsGameObjectHandle.cpp:51

Stack trace: 
0) editor/bin/x64/Debug/libbsf.so: bs::CrashHandler::getStackTrace[abi:cxx11]()+0x59 [0x7ff7dc23cc19]
1) editor/bin/x64/Debug/libbsf.so: bs::CrashHandler::logErrorAndStackTrace(std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, unsigned int) const+0x1b0 [0x7ff7dc226440]
2) editor/bin/x64/Debug/libbsf.so: bs::CrashHandler::reportCrash(std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, bs::StdAlloc<char, bs::GenAlloc> > const&, unsigned int) const+0x4f [0x7ff7dc23c7ef]
3) editor/bin/x64/Debug/libbsf.so: bs::GameObjectHandleBase::throwIfDestroyed() const+0xca [0x7ff7dc85e1da]
4) editor/bin/x64/Debug/libEditorCore.so: bs::GameObjectHandle<bs::Component>::get() const+0x1f [0x7ff7ddcbbc1f]
5) editor/bin/x64/Debug/libEditorCore.so: bs::GameObjectHandle<bs::Component>::operator->() const+0x15 [0x7ff7ddcbb225]
6) editor/bin/x64/Debug/libbsf.so: bs::SceneManager::processStateChanges()+0x91 [0x7ff7dc896641]
7) editor/bin/x64/Debug/libbsf.so: bs::SceneManager::_fixedUpdate()+0x29 [0x7ff7dc8970b9]
8) editor/bin/x64/Debug/libbsf.so: bs::CoreApplication::runMainLoop()+0x29c [0x7ff7dc82f54c]
9) editor/bin/x64/Debug/Banshee3D() [0x403216]
10) /usr/lib/libc.so.6: __libc_start_main+0xeb [0x7ff7d8da906b]
11) editor/bin/x64/Debug/Banshee3D() [0x40313a]
```
Thanks for the answer.

**Comment by BearishSun** [comment]
It is almost certainly caused by those new changes - I haven't had a chance to test them with the editor yet.

My best guess is that's caused by a scene object or a component with the `immediate` flag on. Normally their destruction is delayed until the end of frame, letting systems still access them but when that flag is provided the object is destroyed immediately. It's likely `SceneManager` is not properly notified of this fact and keeps a reference to the destroyed object.

**Comment by athanase** [comment]
Ok, that's right. When closing the `GUIWidget` the method `_notifyComponentDestroyed` gets called properly. But then when the method `processStateChanges()` executes the component no longer exists, hence the error.
```
		const ComponentHandle& component = entry.obj;
		const bool alwaysRun = component->hasFlag( ComponentFlag::AlwaysRun );
```

**Comment by BearishSun** [comment]
I haven't had a chance to test this exact issue yet, but I've fixed a (very likely) related issue https://github.com/GameFoundry/bsfExamples/issues/12. 

In case you get a chance to test, let me know if that also resolves your issue.

The fix is in 440391ccb2cefb7d862beaa37004582a2933189a.

**Comment by athanase** [comment]
That fixed the bug for me too. Good work !
